File: itmo/M4236/gostevskiy/2/LockFreePQ/src/main/java/LockFreePriorityQueue.java
import java.util.AbstractQueue;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class LockFreePriorityQueue<E extends Comparable<E>> extends AbstractQueue<E> implements PriorityQueue<E> {

    private Node<E> head;
    private Node<E> tail;

    LockFreePriorityQueue() {
        head = new Node<>(null);
        tail = new Node<>(null);
        head.next = new AtomicMarkableReference<>(tail, false);
    }

    public int size() {
        return 0;
    }

    public boolean isEmpty() {
        return peek() == null;
    }

    public Iterator<E> iterator() {
        throw new UnsupportedOperationException();
    }

    public boolean offer(E e) {
        while (true) {
            PairNode<E> nodes = find(e);
            Node<E> pred = nodes.first;
            Node<E> cur = nodes.second;
            Node<E> node = new Node<>(e);
            node.next.set(cur, false);
            if (pred.next.compareAndSet(cur, node, false, false)) return true;
        }
    }

    public PairNode<E> find(E e) {
        retry: while (true) {
            Node<E> pred = head;
            Node<E> cur = pred.next.getReference();
            if (cur == null || cur == tail) return new PairNode<>(pred, cur);
            Node<E> suc;
            while (true) {
                if (cur == null || cur == tail) return new PairNode<>(pred, cur);
                AtomicMarkableReference<Node<E>> suc_t = cur.next;
                suc = suc_t.getReference();
                boolean mrk = suc_t.isMarked();
                if (mrk) {
                    if (!pred.next.compareAndSet(cur, suc, false, false)) continue retry;
                    cur = suc;
                } else {
                    if (e == null || cur.elem.compareTo(e) >= 0) return new PairNode<>(pred, cur);
                    pred = cur;
                    cur = suc;
                }
            }
        }
    }

    public E poll() {
        while (true) {
            PairNode<E> nodes = find(null);
            Node<E> pred = nodes.first;
            Node<E> cur = nodes.second;
            if (cur == tail) return null;
            Node<E> succ = cur.next.getReference();
            if (!cur.next.compareAndSet(succ, succ, false, true)) continue;
            pred.next.compareAndSet(cur, succ, false, false);
            return cur.elem;
        }
    }

    public E peek() {
        PairNode<E> nodes = find(null);
        Node<E> cur = nodes.second;
        if (cur == tail) return null;
        return cur.elem;
    }

    private static class Node<E> {
        private E elem;
        private AtomicMarkableReference<Node<E>> next;
        private Node(E val) {
            elem = val;
            next = new AtomicMarkableReference<>(null, false);
        }
    }
    private static class PairNode<E> {
        private Node<E> first;
        private Node<E> second;
        private PairNode(Node<E> first, Node<E> second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: itmo/M4236/Kalinin/lab2/src/main/kotlin/itmo2018/se/solution/LockFreePriorityQueue.kt
package itmo2018.se.solution

import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicMarkableReference
import java.util.concurrent.atomic.AtomicReference

class LockFreePriorityQueue<T : Comparable<T>> : PriorityQueue<T>, AbstractQueue<T>() {
    private val fakeTail = Node<T>()
    private val fakeHead = Node(next = fakeTail)
    private var head = AtomicReference<Node<T>>(fakeHead)
    private val atomic_size = AtomicInteger(0)

    override fun offer(key: T): Boolean {
        while (true) {
            val foundNodes = find(key)
            val pred = foundNodes.first
            val curr = foundNodes.second
            val newNode = Node(key, curr)
            if (pred.nextAndMark.compareAndSet(curr, newNode, false, false)) {
                atomic_size.incrementAndGet()
                return true
            }

        }
    }

    override fun iterator(): MutableIterator<T> {
        throw UnsupportedOperationException()
    }

    override fun peek(): T? {
        while (true) {
            val first = fakeHead
            val curr = first.nextAndMark.reference

            if (curr == fakeTail) {
                return null
            } else {
                val cmk = curr.nextAndMark.isMarked
                val succ = curr.nextAndMark.reference
                if (cmk) {
                    first.nextAndMark.compareAndSet(curr, succ, false, false)
                } else {
                    return curr.key!!
                }
            }

        }
    }

    override fun poll(): T {
        while (true) {
            val first = fakeHead
            val curr = first.nextAndMark.reference

            if (curr == fakeTail) {
                throw NoSuchElementException()
            } else {
                val cmk = curr.nextAndMark.isMarked
                val succ = curr.nextAndMark.reference
                if (cmk) {
                    first.nextAndMark.compareAndSet(curr, succ, false, false)
                } else {
                    if (!curr.nextAndMark.compareAndSet(succ, succ, false, true)) {
                        continue
                    }
                    first.nextAndMark.compareAndSet(curr, succ, false, false)
                    atomic_size.decrementAndGet()
                    return curr.key!!
                }
            }

        }
    }

    override val size: Int
        get() = atomic_size.get()

    override fun isEmpty(): Boolean = peek() == null

    private class Node<T>(var key: T? = null, next: Node<T>? = null, marked: Boolean = false) {
        val nextAndMark = AtomicMarkableReference<Node<T>>(next, marked)
    }

    private fun find(key: T): Pair<Node<T>, Node<T>> {
        retry@ while (true) {
            var pred = head.get()
            var curr = pred.nextAndMark.reference
            var succ: Node<T>?
            while (true) {
                succ = curr.nextAndMark.reference
                val cmk = curr.nextAndMark.isMarked
                if (cmk) {
                    if (!pred.nextAndMark.compareAndSet(curr, succ, false, false)) {
                        continue@retry
                    }
                } else {
                    if (curr == fakeTail || curr.key!! >= key) {
                        return Pair(pred, curr)
                    }
                    pred = curr
                    curr = succ
                }
            }
        }
    }
}